### ISSUE TITLE
TestTaintBasedEvictions is flaky

### ISSUE THREAD

**Issue by alenkacz** [kubernetes/kubernetes#83321]
**Which jobs are failing**:
https://testgrid.k8s.io/sig-release-master-blocking#integration-master

**Which test(s) are failing**:
TaintBasedEvictions

**Since when has it been failing**:
probably since it was added in this PR? https://github.com/kubernetes/kubernetes/pull/81856

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-blocking#integration-master

**Reason for failure**:
```
I0930 12:37:26.944]     --- FAIL: TestTaintBasedEvictions/Taint_based_evictions_for_NodeNotReady_with_no_pod_tolerations (35.04s)
I0930 12:37:26.944]         taint_test.go:770: Failed to taint node in test 1 <node-0>, err: timed out waiting for the condition
```

**Anything else we need to know**:
/sig scheduling
/kind flake
/priority important-soon
/milestone v1.17

**Comment by alenkacz** [comment]
cc. @damemi

**Comment by damemi** [comment]
Looking into this. We moved it to integration because it was flaking in e2e too

**Comment by damemi** [comment]
Not sure yet, but I do see this in the logs of the failing tests: `node_lifecycle_controller.go:1117] Controller detected that all Nodes are not-Ready. Entering master disruption mode.` which, if I understand this, would remove all the taints and mark nodes as reachable (which would explain why the test is failing waiting for node taints: https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/taint_test.go#L769). But I need to figure out why all the nodes are becoming unavailable

**Comment by alenkacz** [comment]
@liggitt is that a full fix for this flake or just partial? I still saw it failed yesterday so just checking... :)

**Comment by lentzi90** [comment]
I think I also run into this on 13th Oct:
```
I1013 12:04:38.321]     --- FAIL: TestTaintBasedEvictions/Taint_based_evictions_for_NodeNotReady_and_200_tolerationseconds (35.07s)
I1013 12:04:38.321]         taint_test.go:770: Failed to taint node in test 0 <node-1>, err: timed out waiting for the condition
```
Full logs:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/78623/pull-kubernetes-integration/1183345872050589697/build-log.txt

**Comment by Huang-Wei** [comment]
Now #84036 gets merged, which _might_ fix the flake. Let's keep an eye on this.

**Comment by Huang-Wei** [comment]
@tedyu reported it still flaked: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1184738536776011777

@damemi Can you continue looking into this?

**Comment by damemi** [comment]
@Huang-Wei I'll keep trying, from my previous debugging it seems that in some cases the test cluster is entering full disruption mode and removing all of the taints from the nodes. I know we specifically made changes to the test (using multiple nodes) to ensure it wouldn't become fully disrupted, do you have any idea what could still be causing that?

**Comment by Huang-Wei** [comment]
@damemi There is some flag controlling how the dispution thing is managed internally, maybe you can check how a node lifecycle manager is constructed.

**Comment by draveness** [comment]
The test still flakes after we merge #84036 
 https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82703/pull-kubernetes-integration/1185364587571056642/

**Comment by Huang-Wei** [comment]
Another flake: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/83578/pull-kubernetes-integration/1190306355727044611/

**Comment by damemi** [comment]
These flakes are consistently (always) timeouts waiting on node taints:
```
--- FAIL: TestTaintBasedEvictions/Taint_based_evictions_for_NodeNotReady_with_no_pod_tolerations (34.97s)
        taint_test.go:770: Failed to taint node in test 1 <node-2>, err: timed out waiting for the condition
```

https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/taint_test.go#L769-L771

**Comment by Huang-Wei** [comment]
I'm working on this. Will send out a PR soon.

**Comment by Huang-Wei** [comment]
/reopen

We see another flake since #84766 got merged: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84738/pull-kubernetes-integration/1192120553314455554/.

However, don't panic :) The original "Entering master disruption mode" cause has been fixed - you won't see that message in the above log. This time it's a smaller issue. I will explain it in another PR.

**Comment by droslean** [comment]
/reopen

The job is still flaky

**Comment by droslean** [comment]
/close 

But for another test :)
